### DISCUSSION TITLE
Stm32F302C8t6 IRQ shenanigan and D+ pull up resistor

### DISCUSSION THREAD

**Discussion by cdg66** [hathach/tinyusb#833] [Q&A] [answered]
I've made a board with an stm32F302C8 and I've notice that the usb IRQ is weird the irq is somewhat bound to the can IRQ and separated in two LP and HP.
My question is where should I call the _tud_int_handler()_;
I'm currently calling in both LowPriority and HighPriority IRQ but it dont show up on my computer. Should i call it in _USBWakeUp_RMP_IRQn()_ instead cause ive noticed this in _dcd_stm32_fsdev.c_? 

> // Some STM32F302/F303 devices allow to remap the USB interrupt vectors from
  // shared USB/CAN IRQs to separate CAN and USB IRQs.
  // This dynamically checks if this remap is active to disable the right IRQs.
  #ifdef SYSCFG_CFGR1_USB_IT_RMP
  if (SYSCFG->CFGR1 & SYSCFG_CFGR1_USB_IT_RMP)
  {
    NVIC_DisableIRQ(USB_HP_IRQn);
    NVIC_DisableIRQ(USB_LP_IRQn);
    NVIC_DisableIRQ(USBWakeUp_RMP_IRQn);
  }
  else
  #endif
  {
    NVIC_DisableIRQ(USB_HP_CAN_TX_IRQn);
    NVIC_DisableIRQ(USB_LP_CAN_RX0_IRQn);
    NVIC_DisableIRQ(USBWakeUp_IRQn);
  }

Thx for your help

**Comment by pigrew** [comment]
I suggest calling `tud_int_handler` from all three interrupt handlers. However, based on reading the reference manual, I think it's only actually required to be called from the LP handler (since I believe that all events trigger the LP interrupt, with some also triggering the other ones).

_(I've never tried remapping the interrupts, so the below is what I understand based on the reference manual, and I have not tried to have different handlers for different interrupts.)_

It doesn't hurt (except for wasting CPU time) to call the interrupt handler too often. The handler checks the interrupt flags, and will do nothing in the case that there is no event flag set.

The manual for your MCU ( RM0365 Rev 8) seems to have an error, it lists the remap bit (SYSCFG_CFGR1.5), but later says it is reserved, so I don't know if it works in the F302C8 or not.

My guess is that there is some other issue causing your enumeration failure.

**Reply by cdg66** [reply]
I'm able to get Interupt for a connection and a disconect and tinyUSB is able to detect it but afer that no more int are triggered.

**Reply by hathach** [reply]
Have you tried to use the stock example first ? It is always good to start with something working first

**Reply by cdg66** [reply]
you mean the stm32 middleweare?

**Reply by hathach** [reply]
No I mean this repo example. You should try it out first, the build instruction can be found in https://github.com/hathach/tinyusb/blob/master/docs/getting_started.md

**Reply by cdg66** [reply]
What ive tried is STM32 Middleweare, UAC2 Headset example and HID Composite. In all case I'm able to get an interupt for a connection and unpulg of the usb and nothing else. What else could be causing that i dont have no more packet. I'm using stm32IDE and cube mx and the config like in the discussion  #633. I've made it working with an stm32F4 discovery board but not on the F3

**Reply by hathach** [reply]
I don't use stm32 cube and middleware, and couldn't help any further. keep debugging, I am sure it is something minor missing somehwere.

**Comment by cdg66** [comment]
[Schematic.PDF](https://github.com/hathach/tinyusb/files/6487584/Schematic.PDF)
here is the schematic. I know that D- and D+ are swap on the stm32 but they have been patched on the board so that's not the problem

**Reply by pigrew** [reply]
Your crystal is higher load capacitance than I've used in the past. The datasheets isn't so clear, but they suggest capacitors of less than 25 pF each. My suggestion is to verify that the oscillator is able to start up and that its frequency is correct. (Enable the MCO output and input that into a frequency counter or oscilloscope).

Otherwise, I didn't notice anything else of concern.

**Reply by cdg66** [reply]
what load capacitance do you go to?

**Reply by pigrew** [reply]
I usually use SMD 8MHz crystals specified with about 10pF load capacitance. With them, I'd use a pair of 16 or 18 pF caps.

I'm not saying it won't work with your crystal, but verifying the oscillator is what I'd suggest checking. Looking at the RCC status registers can tell you if the oscillator is oscillating, but won't tell you if the frequency is correct.

**Reply by cdg66** [reply]
the HSEON bit is on in the RCC_CR register

**Reply by pigrew** [reply]
`RCC_CR.HSEON` is what you set to request the HSE to be enabled. You'd need to look at `RCC_CR.HSERDY`, which should be set to 1 if the hardware thinks the HSE is stable.

**Reply by cdg66** [reply]
RCC_CR.HSERDY is set to 1

**Reply by cdg66** [reply]
should i call __HAL_RCC_USB_CLK_ENABLE() before the init as its stated in 

> void dcd_init (uint8_t rhport)
{
  /* Clocks should already be enabled */
  /* Use __HAL_RCC_USB_CLK_ENABLE(); to enable the clocks before calling this function */

  /* The RM mentions to use a special ordering of PDWN and FRES, but this isn't done in HAL.
   * Here, the RM is followed. */

**Comment by cdg66** [comment]
can you tell me for witch specific mcu in the f3 series you acctually tested the stack? I could try to see any difference

**Reply by hathach** [reply]
https://github.com/hathach/tinyusb/blob/master/docs/boards.md

**Comment by cdg66** [comment, marked as answer]
I think i found my problem  in https://www.st.com/content/ccc/resource/technical/document/application_note/group0/0b/10/63/76/87/7a/47/4b/DM00296349/files/DM00296349.pdf/jcr:content/translations/en.DM00296349.pdf the f2 fammily need a 1.5k pullup to D+ witch i didnt added and were the discovery board do have I'll add one monday and keep you up to date thx for your help.

**Reply by cdg66** [reply]
After a quick test it confirm that i need a pull up resistor to 3.3v in order to tell the host that im a full speed device otherwise it think that there is nothing pluged in(F3 Don't have it integrated). Thank alot for your help and your quick response.

**Reply by pigrew** [reply]
Good to hear you figured it out. I had forgotten about the pull-ups.
